File: cpp/0212.cpp
#include <vector>
#include <string>
#include <iostream>

using namespace std;

class TrieNode {
public:
    TrieNode* child[26] = {};
    string word = "";
};

class Trie {
public:
    TrieNode* root;
    Trie(const vector<string>& words) {
        root = new TrieNode();
        for (const string& w: words) {
            TrieNode* curr = this->root;
            for (char c: w) {
                int idx = c - 'a';
                if (!curr->child[idx]) {
                    curr->child[idx] = new TrieNode();
                }
                curr = curr->child[idx];
            }
            curr->word = w;
        }
    }
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        Trie* trie = new Trie(words);
        int m = board.size();
        int n = board[0].size();
        vector<string> res;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                dfs(i, j, board, trie->root, res);
            }
        }
        return res;
    }

    // check the board[i][j] is valid and belongs to the trie
    void dfs(int i, int j, vector<vector<char>>& board, TrieNode* trie, vector<string>& res) {
        int m = board.size();
        int n = board[0].size();
        if (i < 0 || j < 0 || i >= m || j >= n || board[i][j] == '#') return;
        TrieNode* t = trie->child[board[i][j] - 'a'];
        if (!t) return;
        if (t->word != "") {
            res.push_back(t->word);
            // change the word state to avoid the rework
            t->word = "";
        }
        const char c = board[i][j];
        board[i][j] = '#';
        dfs(i+1, j, board, t, res);
        dfs(i, j+1, board, t, res);
        dfs(i-1, j, board, t, res);
        dfs(i, j-1, board, t, res);
        board[i][j] = c;
    }
};

int main() {
    Solution sol;
    // vector<vector<char>> board = {
    //     {'o','a','a','n'},
    //     {'e','t','a','e'},
    //     {'i','h','k','r'},
    //     {'i','f','l','v'}
    // };
    // vector<string> words = {"oath","pea","eat","rain"};
    // vector<vector<char>> board = {{'a','b'},{'c','d'}};
    // vector<string> words = {"abcb"};
    vector<vector<char>> board = {
        {'o','a','b','n'},
        {'o','t','a','e'},
        {'a','h','k','r'},
        {'a','f','l','v'}
    };
    vector<string> words = {"oa","oaa"};
    for (auto w: sol.findWords(board, words)) {
        cout << w << endl;
    }
    return 0;
}

File: python/0212.py
from collections import defaultdict
from typing import List

class TrieNode:
    def __init__(self):
        self.next = defaultdict(TrieNode)
        self.word = ''

class Trie:
    def __init__(self, words):
        self.root = TrieNode()
        for w in words:
            node = self.root
            for c in w:
                node = node.next[c]
            node.word = w


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        out = []
        trie = Trie(words)
        m, n = len(board), len(board[0])

        def dfs(node, i, j):
            if not (0 <= i < m and 0 <= j < n and board[i][j] in node.next): return

            node = node.next[board[i][j]]
            if node.word:
                out.append(node.word)
                # remember to clean
                node.word = ""

            tmp, board[i][j] = board[i][j], '#'
            dfs(node, i+1, j)
            dfs(node, i-1, j)
            dfs(node, i, j+1)
            dfs(node, i, j-1)
            board[i][j] = tmp

        for i in range(m):
            for j in range(n):
                dfs(trie.root, i, j)

        return out




board = [
    ["o","a","a","n"],
    ["e","t","a","e"],
    ["i","h","k","r"],
    ["i","f","l","v"]
]
words = ["oath","pea","eat","rain"]
print(Solution().findWords(board, words))
